clients/src/main/java/org/apache/kafka/common/requests/LeaderAndIsrRequest.java

```diff
@@ -18,290 +18,132 @@
-        private final Set<Node> liveLeaders;
+        private final LeaderAndIsrRequestData data;
 
         public Builder(short version, int controllerId, int controllerEpoch, long brokerEpoch,
-                       Map<TopicPartition, PartitionState> partitionStates, Set<Node> liveLeaders) {
+                       Map<TopicPartition, LeaderAndIsrRequestPartition> partitionStates, Set<Node> liveLeaders) {
             super(ApiKeys.LEADER_AND_ISR, version, controllerId, controllerEpoch, brokerEpoch);
-            this.partitionStates = partitionStates;
-            this.liveLeaders = liveLeaders;
+
+            List<LeaderAndIsrLiveLeader> leaders = liveLeaders.stream().map(n -> new LeaderAndIsrLiveLeader()
+                    .setBrokerId(n.id())
+                    .setHostName(n.host())
+                    .setPort(n.port())
+            ).collect(Collectors.toList());
+
+            LeaderAndIsrRequestData data = new LeaderAndIsrRequestData();
+            data
+                .setControllerId(controllerId)
+                .setControllerEpoch(controllerEpoch)
+                .setBrokerEpoch(brokerEpoch)
+                .setLiveLeaders(leaders);
+
+            if (version >= 2) {
+                Map<String, Map<Integer, LeaderAndIsrRequestPartition>> groupedMap = CollectionUtils.groupPartitionDataByTopic(partitionStates);
+                List<LeaderAndIsrRequestTopicState> topicStates = groupedMap.entrySet().stream().map(entry ->
+                    new LeaderAndIsrRequestTopicState()
+                        .setName(entry.getKey())
+                        .setPartitionStatesV0(new ArrayList<>(entry.getValue().values()))
+                ).collect(Collectors.toList());
+                data.setTopicStates(topicStates);
+            } else {
+                data.setPartitionStatesV0(new ArrayList<>(partitionStates.values()));
+            }
+
+            this.data = data;
         }
 
         @Override
         public LeaderAndIsrRequest build(short version) {
-            return new LeaderAndIsrRequest(controllerId, controllerEpoch, brokerEpoch, partitionStates, liveLeaders, version);
+            return new LeaderAndIsrRequest(data, version);
         }
 
         @Override
         public String toString() {
-            StringBuilder bld = new StringBuilder();
-            bld.append("(type=LeaderAndIsRequest")
-                .append(", controllerId=").append(controllerId)
-                .append(", controllerEpoch=").append(controllerEpoch)
-                .append(", brokerEpoch=").append(brokerEpoch)
-                .append(", partitionStates=").append(partitionStates)
-                .append(", liveLeaders=(").append(Utils.join(liveLeaders, ", ")).append(")")
-                .append(")");
-            return bld.toString();
+            return data.toString();
         }
     }
 
+    private final LeaderAndIsrRequestData data;
     private final Map<TopicPartition, PartitionState> partitionStates;
```

Comment:
We should be able to use LeaderAndIsrRequestData directly rather than copying it all here (I can understand if you want to do that in a follow-on...)